spopt/locate/p_center.py

```diff
@@ -0,0 +1,216 @@
+        cost_matrix: np.array
+            two-dimensional distance array between facility points and demand point
+        weights: np.array
+            one-dimensional service load or population demand
+        p_facilities: int
+            number of facilities to be located
+        name: str, default="p-center"
+            name of the problem
+
+        Returns
+        -------
+        PCenter object
+        """
+        r_cli = range(cost_matrix.shape[0])
+        r_fac = range(cost_matrix.shape[1])
+
+        model = pulp.LpProblem(name, pulp.LpMinimize)
+
+        weights = np.reshape(weights, (cost_matrix.shape[0], 1))
+        aij = weights * cost_matrix
+
+        p_center = PCenter(name, model, aij)
+
+        FacilityModelBuilder.add_facility_integer_variable(p_center, r_fac, "y[{i}]")
+        FacilityModelBuilder.add_client_assign_integer_variable(
+            p_center, r_cli, r_fac, "z[{i}_{j}]"
+        )
+        FacilityModelBuilder.add_weight_continuous_variable(p_center)
+
+        p_center.__add_obj()
+
+        FacilityModelBuilder.add_facility_constraint(
+            p_center, p_center.problem, p_facilities
+        )
+        FacilityModelBuilder.add_assignment_constraint(
+            p_center, p_center.problem, r_fac, r_cli
+        )
+        FacilityModelBuilder.add_opening_constraint(
+            p_center, p_center.problem, r_fac, r_cli
+        )
+        FacilityModelBuilder.add_minimized_maximum_constraint(
+            p_center, p_center.problem, cost_matrix, r_fac, r_cli
+        )
+
+        return p_center
+
+    @classmethod
+    def from_geodataframe(
+        cls,
+        gdf_demand: GeoDataFrame,
+        gdf_fac: GeoDataFrame,
+        demand_col: str,
+        facility_col: str,
+        weights_cols: str,
+        p_facilities: int,
+        distance_metric: str = "euclidean",
+        name: str = "p-center",
+    ):
+        """
+        Create a PCenter object based on geodataframes. Calculate the cost matrix between demand and facility,
+        and then use from_cost_matrix method.
+
+        Parameters
+        ----------
+        gdf_demand: geopandas.GeoDataFrame
+            demand geodataframe with point geometry
+        gdf_fac: geopandas.GeoDataframe
+            facility geodataframe with point geometry
+        demand_col: str
+            demand geometry column name
+        facility_col: str
+            facility candidate sites geometry column name
+        weights_cols: str
+            weight column name representing service load or demand
+        p_facilities: int
+            number of facilities to be located
+        distance_metric: str, default="euclidean"
+            metrics supported by :method: `scipy.spatial.distance.cdist` used for the distance calculations
+        name: str, default="p-median"
+            name of the problem
+
+        Returns
+        -------
+        PCenter object
+        """
+
+        service_load = gdf_demand[weights_cols].to_numpy()
```

Comment:
We may want to keep this in if we will be adding capacity constraints, but the weighted distance should not be used in the *p*-center problem.

Reply:
Hm. I think that I did not delete this LOC because in constraints I used the cost_matrix without weighting. I ran all the tests without this LOC and gave me no errors.

Reply:
Though it seems in LOC177 `service_load` is being passed into `from_cost_matrix()`, which is then used in LOC76-77 the calculate `aij` (the population weighted distance matrix). Then `aij` is passed in to `PCenter()` in LOC 79. Since the *p*-center problem should be solved without population demand, only the unweighted distances should be used as an input. Best to remove all mention of `aij` & weights from the `PCenter` class.